Eth tests logging

cabrador opened the pull request:
## Description

This PR improves logging in `eth-tests`
1) readable errors
2) overall test count

## Type of change

- [ ] New feature (non-breaking change which adds functionality)

cabrador committed:
Improve eth-test logging

cabrador committed:
Improve error gathering

cabrador committed:
Merge branch 'develop' into c/eth-tests-logging

wsodsong commented on executor/extension/logger/error_logger.go:
```diff
@@ -83,7 +83,10 @@ func (l *errorLogger[T]) PostRun(_ executor.State[T], ctx *executor.Context, _ e
 	}
 
 	if len(l.errors) != 0 {
-		return fmt.Errorf("total %v errors occurred: %v", len(l.errors), errors.Join(l.errors...))
```
I feel this line is still needed to know the total number of errors.

cabrador replied:
This change was moved to https://github.com/Fantom-foundation/Aida/pull/1191

Each error has number assigned to it. So we know how many errors we have.

wsodsong reviewed (commented):
It would be nice to see an example of the new error message.

cabrador committed:
Merge branch 'develop' into c/eth-tests-logging

cabrador committed:
Revert error_logger.go changes.

HerbertJordan commented on executor/extension/validator/eth_state_test_validator.go:
```diff
@@ -66,10 +64,10 @@ func (e *ethStateTestValidator) PostTransaction(state executor.State[txcontext.T
 	}
 	if got == nil && want != nil {
-		err = fmt.Errorf("expected error %w, got no error\nTest info:\n%s", want, state.Data)
+		err = fmt.Errorf("unexpected error %w, got no error\ntest-info: %s", want, state.Data)
```
I think `expected error` was correct ...

HerbertJordan commented on ethtest/test_case_splitter.go:
```diff
@@ -113,7 +113,7 @@ func (s *TestCaseSplitter) SplitStateTests() (dividedTests []Transaction, err er
 					fork = "Merge"
 				}
-				txCtx := newStateTestTxContext(stJson, msg, post, chainCfg, fork, postNumber)
+				txCtx := newStateTestTxContext(stJson, msg, post, chainCfg, fork, overall)
```
The function `newStateTestTxContext` expects the post-number as its last argument and the resulting struct will store the given number as its `postNumber`.

Not sure what is the benefit of not using the `postNumber` here.

Assigning all tests a global number has conceptional issues:
- it makes the numbering dependent on the order in which files are enumerated by the list of json files
- adding, renaming, or removing a file will change the number of a test
- adding or removing post-cases of any test will change the number of subsequent tests

So the test number is not a very reliable bit of information, and it could not be used for e.g. filtering specific tests.

What is the motivation for this change?

cabrador replied:
First thought when I implemented this was to number each `post` within `fork` within `file`
ie.
Test with path `/random/path/` contained two forks `Shangai` and `Cancun`. Each fork contained two `posts` (test cases).

So the intention was to find exactly the test that is failing using this number + path + fork.
Although this does not really work since there are tests that contain more tests within a file such as [this one](https://github.com/ethereum/tests/blob/develop/GeneralStateTests/Pyspecs/cancun/eip1153_tstore/reentrant_selfdestructing_call.json).

Adding the overall number helps understanding how many tests has been produced so far.

There is a `generatedTestHash` property for each test within the file which would distinguish each test within a file. We can add that but I feel like thats overcomplicating the logging? I believe knowing which file + which fork failed is enough. Wdyt?

HerbertJordan replied:
Sounds like the tuple `(path, test, post, fork)` where `test` is the index of the test in the file (which seems to be a better choice than the [name](https://github.com/ethereum/tests/blob/9201075490807f58811078e9bb5ec895b4ac01a5/GeneralStateTests/Pyspecs/cancun/eip1153_tstore/reentrant_selfdestructing_call.json#L2)) would be able to uniquely identify a test, independent of the content of any other test file.

This order would correspond to the navigation path:
- first find the file
- within the file, find the test
- within the test, find the post block
- within the post block, find the fork

cabrador replied:
The JSON converts into a map. We cannot guarantee the order of the tests.

HerbertJordan replied:
Sounds like sorting them by their key (which I guess is a string) would do the trick.

Otherwise, just use the string key.

cabrador committed:
Fix error message.

cabrador committed:
Merge branch 'develop' into c/eth-tests-logging

cabrador committed:
Merge branch 'develop' into c/eth-tests-logging

HerbertJordan reviewed (dismissed):
I do not think that the global numbering is too useful in identifying the failing test case. But I guess it worked so far, by just identifying the file containing the failing test. And as I understand, that is also the granularity in which `aida-vm-sdb` allows the user to define the test target. So it seems fine for me.

cabrador committed:
Add label of the test

cabrador commented:
@HerbertJordan 
I removed the total number and brought back the post number within one test.
I also added the test label to the logging so that we know which test from the file actually failed.

So the log looks like this:
<img width="564" alt="Screenshot 2024-10-08 at 9 05 11" src="https://github.com/user-attachments/assets/c7cde701-6281-4e29-8b34-cc0cf7cd283f">

And the error message looks like this:
<img width="562" alt="Screenshot 2024-10-08 at 9 07 05" src="https://github.com/user-attachments/assets/c11c769b-868e-4a50-97f9-404b8b68d81e">

Used [test](https://github.com/ethereum/tests/blob/9201075490807f58811078e9bb5ec895b4ac01a5/EIPTests/StateTests/stExample/blobtxExample.json) in screenshots


I guess this is the best iteration so far.

HerbertJordan commented on ethtest/state_test_context.go:
```diff
@@ -79,5 +87,9 @@ func (s *stateTestContext) GetResult() txcontext.Result {
+	return fmt.Sprintf(
+		"Test path: %v\n"+
+			"Test label: %v\n"+
+			"Fork: %v\n"+
+			"PostNumber: %v\n", s.path, s.testLabel, s.fork, s.postNumber)
```
nit: since in the JSON files the hierarchy is Post > Fork, these should be printed in the reverse order.

cabrador replied:
good point

HerbertJordan reviewed (dismissed):
Nice, thanks!

cabrador committed:
Swap fork and post.

cabrador committed:
Merge branch 'develop' into c/eth-tests-logging

cabrador committed:
Merge branch 'develop' into c/eth-tests-logging

# Conflicts:
#	ethtest/state_test_context.go
#	executor/extension/validator/eth_state_test_validator.go
